File: leetcode_python/124.Binary_Tree_Maximum_Path_Sum.py
#!/usr/bin/env python3
# coding: utf-8

# Time complexity: O()
# Space complexity: O()

# https://leetcode.com/problems/binary-tree-maximum-path-sum/


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxPathSum(self, root):
        """
        path sum = max(0, sum(root.left)) + root.val + max(0, sum(root.right))
        
        cur node works as a child node and return
            cur .val + max(0, sum(cur.left), sum(cur.right))


           -10
           / \
          9  20
            /  \
           15   7   


           -10
           / \
          9  42  

        cur_max = 42 | 41
        
        test case &corner case:
            1
            
            
        time O(n)
        space O(1)
        """
        self.max_val = float('-inf')
        self.dfs(root)
        return self.max_val
        
    def dfs(self, root):
        if not root:
            return 0
        
        left = self.dfs(root.left)
        right = self.dfs(root.right)
        
        cur = root.val + left + right
        self.max_val = max(self.max_val, cur)
        
        return max(0, root.val + max(left, right))


# without global valuable
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        res = self.dfs(root)
        return res[1]
        
    def dfs(self, root):
        if not root:
            return 0, float('-inf')
        
        left, left_max = self.dfs(root.left)
        right, right_max = self.dfs(root.right)
        
        cur = root.val + left + right
        cur_max = max(left_max, right_max, cur)
        
        return max(0, root.val + max(left, right)), cur_max

File: leetcode_go/124.Binary_Tree_Maximum_Path_Sum.go
import "math"

// https://leetcode.com/problems/binary-tree-maximum-path-sum/

func maxPathSum(root *TreeNode) int {
	if root == nil {
		return 0
	}

	maxVal := math.MinInt32

	dfs(root, &maxVal)
	return maxVal
}

func dfs(node *TreeNode, val *int) int {
	if node == nil {
		return 0
	}

	left := dfs(node.Left, val)
	right := dfs(node.Right, val)
	*val = max(*val, node.Val+left+right)

	return max(0, node.Val+max(left, right))
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}